hacks/farmer.js

```diff
@@ -0,0 +1,135 @@
+const fs = require("graceful-fs");
+const createHashFromString = require("./hash");
+
+/**
+ * Hash map of where the parserOptions are stored.  We'll generate a hash from the parserOptions,
+ * then use that hash in this key/value store.
+ * key - The hash of the options, value - The location the actual options are stored.
+ */
+const parserOptionsMap = {};
+tmp.setGracefulCleanup();
+
+var farm;
+var _buildModule;
+
+function startFarm() {
+  if(farm) throw new Error("farm already started");
+  farm = workerFarm({
+    autoStart: true,
+    maxConcurrentCallsPerWorker: Infinity,
+    maxConcurrentWorkers: os.cpus().length - 1,
+    maxRetries: 2, // Allow for a couple of transient errors.
+  },
+   require.resolve("./worker"),
+   ["buildModule"]
+  );
+  _buildModule = pify(farm.buildModule);
+}
+
+function stopFarm() {
+  workerFarm.end(farm);
+  farm = undefined;
+  _buildModule = undefined;
+}
+
+module.exports = {
+  startFarm,
+  stopFarm,
+  buildModule(module, resolver, options, compilation, inputFS, callback) {
+    if(!logStarted && console.profile) {
+      console.profile(`build module ${logStarted}`);
+    }
+    module.cacheable = false; /* womp */
+    const loaderContext = module.createLoaderContext(resolver, options, compilation, inputFS);
+    /** gross, but none of this is serializable **/
+    loaderContext.emitWarning = undefined;
+    loaderContext.emitError = undefined;
+    loaderContext.exec = undefined;
+    loaderContext.resolve = undefined;
+    loaderContext.resolveSync = undefined;
+    loaderContext.emitFile = undefined;
+    loaderContext.emitFileSync = undefined;
+    loaderContext.options = undefined;
+    loaderContext._module = undefined;
+    loaderContext._compilation = undefined;
+    loaderContext._compiler = undefined;
+    loaderContext.fs = undefined;
+    loaderContext.loadModule = undefined;
+    loaderContext.readResource = undefined;
+
+
+    /**
+     * Parser options (plugins) are large, so we'll write them to disk and
+     * pass a reference instead of passing the entire object to each worker
+     * call. This will let workers cache an instance of a parser using these
+     * options, and will cut down the overhead of ipc.
+     */
+    const serializedModule = module.serialize();
+    const parserProps = serializedModule.parser;
+    serializedModule.parser = undefined; // pull this of off the parser.
+    const stringifiedParserProps = JSON.stringify(parserProps);
+    const parserHash = createHashFromString(stringifiedParserProps);
+    if(!parserOptionsMap[parserHash]) {
+      const parserOptionsLocation = tmp.fileSync({ prefix: "parser-" }).name;
+      fs.writeFileSync(parserOptionsLocation, stringifiedParserProps);
+      parserOptionsMap[parserHash] = parserOptionsLocation;
+    }
+
+    const parserOptionsLocation = parserOptionsMap[parserHash];
+
+    let opts = {
+      moduleProps: serializedModule,
+      loaderContext,
+      // compilation, womp.
+      compilation: {},
```

Comment:
I'm almost 100% positive that passing a `{}` in place of the actual compilation object to `Parser.parse` will break *something* for *somebody*, but I'm also pretty sure that the entire object isn't needed.

I mentioned type checking in Slack, but this is another example of where it would be helpful.  I'd like to be able to confidently build a serializable object that exactly matches only what `Parser.parse` needs.

Reply:
@sokra this bit we should maybe assist @gdborton on